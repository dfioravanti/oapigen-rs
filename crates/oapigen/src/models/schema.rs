use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use std::hash::Hash;
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// SchemaAsRust represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct SchemaAsRust {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The macros that will be applied to this type as a TokenStream
    pub tokenized_macros: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,
    /// The optional comment to the schema
    pub comment: Option<String>,

    pub current_type: CurrentType,
}
impl Eq for SchemaAsRust {}
impl PartialEq for SchemaAsRust {
    fn eq(&self, other: &Self) -> bool {
        self.to_token_stream().to_string() == other.to_token_stream().to_string()
    }
}

impl Hash for SchemaAsRust {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.to_token_stream().to_string().hash(state);
    }
}

impl fmt::Display for SchemaAsRust {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .fold(TokenStream::new(), |mut acc, (_, v)| {
                acc.extend(v.clone());
                acc
            });

        let output = if imports.is_empty() {
            format!("{}", self.to_token_stream())
        } else {
            format!("{} \n {}", imports, self.to_token_stream())
        };

        let b = match syn::parse_file(&output.to_string()) {
            Ok(b) => b,
            Err(e) => panic!("{}", e),
        };
        let formatted = prettyplease::unparse(&b);

        write!(f, "{}", formatted)
    }
}

impl ToTokens for SchemaAsRust {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let generated_tokens = match self.current_type {
            CurrentType::Type => tokenize_type(self),
            CurrentType::Const => todo!(),
            CurrentType::Enum => todo!(),
            CurrentType::Vec => todo!(),
            CurrentType::Struct => todo!(),
        };

        tokens.clone_from(&generated_tokens);
    }
}

fn tokenize_type(tokenized_schema: &SchemaAsRust) -> TokenStream {
    let tokenized_name = &tokenized_schema.tokenized_name;
    let tokenized_type = &tokenized_schema.tokenized_type;
    let tokenized_macros = &tokenized_schema.tokenized_macros;
    let tokenized_comment = match &tokenized_schema.comment {
        None => TokenStream::new(),
        Some(c) => {
            quote! { #[doc = #c] }
        }
    };

    quote! {
        #tokenized_comment
        #tokenized_macros
        struct #tokenized_name(#tokenized_type);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime; },
        );

        let parsed_schema = SchemaAsRust {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            tokenized_macros: quote! { #[derive(Serialize, Deserialize, Debug)] },
            imports,
            comment: Some("My favourite comment".to_string()),
            current_type: CurrentType::Type,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_snapshot!(formatted_parsed_schema);
    }
}
