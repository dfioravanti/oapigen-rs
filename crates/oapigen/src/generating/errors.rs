use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum GeneratingError {
    #[error("could not write to file.")]
    WriteToFileError(#[from] io::Error),
}
