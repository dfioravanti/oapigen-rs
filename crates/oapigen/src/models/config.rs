use std::path::PathBuf;

pub enum DateTimeLibraries {
    Chrono,
    Jiff,
}

pub struct Libraries {
    pub datetime: DateTimeLibraries,
}

impl Default for Libraries {
    fn default() -> Self {
        Libraries {
            datetime: DateTimeLibraries::Chrono,
        }
    }
}

#[derive(Default)]
pub struct Config {
    pub output_path: PathBuf,
    pub libraries: Libraries,
}
