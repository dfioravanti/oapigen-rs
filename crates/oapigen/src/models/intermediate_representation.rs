#[derive(Debug, serde::Serialize)]
pub struct Struct {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<Field>,
}

#[derive(Debug, serde::Serialize)]
pub struct Field {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub is_required: bool,
}

pub struct ConstantField {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub value: serde_json::Value,
}
