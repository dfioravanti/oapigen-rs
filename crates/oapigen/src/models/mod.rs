pub mod intermediate_representation;
pub mod schema;

pub use intermediate_representation::*;
pub use schema::*;
