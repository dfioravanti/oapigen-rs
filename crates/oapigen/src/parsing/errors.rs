use oas3::spec;
use proc_macro2::LexError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParsingError {
    #[error("could not parse the string to TokenStream.")]
    CannotParseString(#[from] LexError),
    #[error("could not resolve the reference with the passed spec.")]
    ReferenceNotFound(#[from] spec::RefError),
}
