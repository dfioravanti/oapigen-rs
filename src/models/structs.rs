use proc_macro2::TokenStream;
use quote::{ToTokens, TokenStreamExt, quote};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// TokenizedType represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct TokenizedSchema {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,

    pub current_type: CurrentType,
}

impl fmt::Display for TokenizedSchema {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .map(|(_, v)| v.to_string())
            .collect::<Vec<_>>();

        write!(
            f,
            "imports => {:?} \n schema => {:?}",
            imports,
            self.to_token_stream().to_string()
        )
    }
}

impl quote::ToTokens for TokenizedSchema {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let tokenized_name = &self.tokenized_name;
        let tokenized_type = &self.tokenized_type;

        let generated_tokens = quote! {

            #tokenized_name: #tokenized_type

        };

        tokens.clone_from(&generated_tokens);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime },
        );

        let parsed_schema = TokenizedSchema {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            imports,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_yaml_snapshot!(formatted_parsed_schema);
    }
}
