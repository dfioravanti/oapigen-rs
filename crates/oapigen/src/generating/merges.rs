use crate::models;
use crate::models::schema;
use proc_macro2::TokenStream;
use quote::ToTokens;
use std::collections::HashSet;

pub fn merge_schemas(schemas: Vec<schema::SchemaAsRust>) -> TokenStream {
    let imports: Vec<_> = schemas.iter().map(|schema| &schema.imports).collect();
    let mut output = merge_imports(imports);

    let unique_types: HashSet<_> = HashSet::from_iter(schemas.iter());
    let types = unique_types.iter().fold(TokenStream::new(), |mut acc, s| {
        acc.extend(s.to_token_stream());
        acc
    });

    output.extend(types);
    output
}

/// merge_imports merges the [TokenStream] that represent the imports used by the models.
/// So for examples if we have two models that use `use chrono::DateTime;` the resulting
/// [TokenStream] will contain only one mention of `use chrono::DateTime;`.
fn merge_imports(all_imports: Vec<&models::Imports>) -> TokenStream {
    let mut seen: HashSet<String> = HashSet::new();
    let mut output = TokenStream::new();
    for imports in all_imports {
        for (import_as_string, actual_import) in imports {
            if !seen.contains(import_as_string) {
                seen.insert(import_as_string.clone());
                output.extend(actual_import.clone().into_iter());
            }
        }
    }

    output
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::CurrentType;
    use quote::quote;

    #[test]
    fn test_merge_imports() {
        let mut imports_1 = models::Imports::new();
        imports_1.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime; },
        );

        let mut imports_2 = models::Imports::new();
        imports_2.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime; },
        );
        imports_2.insert(
            "crate::generating::errors::GeneratingError".to_string(),
            quote! { use crate::generating::errors::GeneratingError; },
        );

        let imports = vec![&imports_1, &imports_2];
        let want = quote! {
            use chrono::DateTime;
            use crate::generating::errors::GeneratingError;
        };

        let got = merge_imports(imports);
        assert_eq!(want.to_string(), got.to_string());
    }

    #[test]
    fn test_merge_schema() {
        let mut imports_1 = models::Imports::new();
        imports_1.insert("chrono".to_string(), quote! { use chrono; });
        let schema1 = schema::SchemaAsRust {
            tokenized_name: quote! { user_time },
            tokenized_type: quote! { chrono::DateTime<chrono::Utc> },
            imports: imports_1,
            current_type: CurrentType::Type,
        };

        let schemas = vec![schema1];
        let got = merge_schemas(schemas);
        let want = quote! {
            use chrono;

            type user_time = chrono::DateTime<chrono::Utc>;
        };
        assert_eq!(want.to_string(), got.to_string());
    }
}
