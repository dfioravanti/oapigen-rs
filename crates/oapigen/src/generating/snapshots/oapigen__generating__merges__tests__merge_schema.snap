---
source: crates/oapigen/src/generating/merges.rs
expression: got.to_string()
---
use chrono ; # [derive (Serialize , Deserialize , Debug)] struct user_time (chrono :: DateTime < chrono :: Utc >) ;
